src/compiler/compile/Component.ts

```diff
@@ -793,20 +793,29 @@ export default class Component {
 				}
 
 				let deep = false;
-				let names: string[] | undefined; 
-
-				if (node.type === 'AssignmentExpression') {
+				let names: string[] = []; 
+
+				if (node.type === 'AssignmentExpression' &&
+					node.left.type === 'ArrayPattern') {
+					walk(node.left, {
```

Comment:
We need to add `deep = true;`.
Now, this code makes compile error but it should not make a compile error.

```svelte
<script>
	const a = 100;
	const arr = [{ a: 1 }, 2];
	[arr[0].a, arr[1] = a] = [arr[1]];
</script>

{JSON.stringify(arr)}
```

Reply:
updated: https://github.com/sveltejs/svelte/pull/7966/commits/28565cfa9947a2db40bf29fdd3f015a6fda863ee